framework/src/neml2/userobjects/NEML2FEInterpolation.C

```diff
@@ -0,0 +1,415 @@
+
+  if (_tid != 0)
+    syncWithMainThread();
+}
+
+void
+NEML2FEInterpolation::invalidateFEMContext()
+{
+  _fem_context_up_to_date = false;
+  invalidateInterpolations();
+}
+
+void
+NEML2FEInterpolation::invalidateInterpolations()
+{
+  _interp_up_to_date = false;
+}
+
+void
+NEML2FEInterpolation::meshChanged()
+{
+  invalidateFEMContext();
+}
+
+void
+NEML2FEInterpolation::initialize()
+{
+  if (_fem_context_up_to_date)
+    return;
+
+  _ndofe.clear();
+  _moose_dof_map.clear();
+  _moose_dof_map_global.clear();
+  _moose_phi.clear();
+  _moose_grad_phi.clear();
+  _local_ndof = 0;
+}
+
+void
+NEML2FEInterpolation::syncWithMainThread()
+{
+  auto & main_uo = _fe_problem.getUserObject<NEML2FEInterpolation>(name(), /*tid=*/0);
+  for (const auto & [var_name, var] : main_uo._moose_vars)
+  {
+    _moose_vars.emplace(var_name, getMOOSEVariable(var_name));
+    if (main_uo._phis.count(var->feType()))
+      getPhi(var_name);
+    if (main_uo._grad_phis.count(var->feType()))
+      getPhiGradient(var_name);
+  }
+}
+
+void
+NEML2FEInterpolation::threadJoin(const UserObject & y)
+{
+  const auto & other = static_cast<const NEML2FEInterpolation &>(y);
+  mooseAssert(_fem_context_up_to_date == other._fem_context_up_to_date,
+              "NEML2FEInterpolation becomes out of sync with other thread");
+
+  if (_fem_context_up_to_date)
+    return;
+
+  auto merge_map_vecs = [](auto & map1, const auto & map2)
+  {
+    for (const auto & [key, map2_val] : map2)
+    {
+      auto & map1_val = map1[key];
+      map1_val.insert(map1_val.end(), map2_val.begin(), map2_val.end());
+    }
+  };
+
+  merge_map_vecs(_moose_dof_map, other._moose_dof_map);
+  merge_map_vecs(_moose_phi, other._moose_phi);
+  merge_map_vecs(_moose_grad_phi, other._moose_grad_phi);
+}
+
+void
+NEML2FEInterpolation::execute()
+{
+  if (_fem_context_up_to_date)
+    return;
+
+  // DOF indices
+  const auto & nl_dof_map = _sys.dofMap();
+  for (const auto & [var_name, var] : _moose_vars)
+  {
+    nl_dof_map.dof_indices(_current_elem, _dof_indices, var->number());
+    auto [it, success] = _ndofe.emplace(var->feType(), _dof_indices.size());
+    if (!success && std::size_t(it->second) != _dof_indices.size())
+      mooseError("DOF map size mismatch for variable ",
+                 var_name,
+                 ", got ",
+                 it->second,
+                 " and ",
+                 _dof_indices.size());
+    auto & moose_dof_map = _moose_dof_map[var_name];
+    auto & moose_dof_map_global = _moose_dof_map_global[var_name];
+    for (auto dof : _dof_indices)
+    {
+      moose_dof_map.push_back(_petsc_solution->map_global_to_local_index(dof));
+      moose_dof_map_global.push_back(dof);
+    }
+  }
+
+  // shape function values
+  for (const auto & [fetype, phi] : _phis)
+  {
+    auto & moose_phi = _moose_phi[fetype];
+    for (auto i : index_range(*phi))
+      for (auto qp : index_range(_q_point))
+        moose_phi.push_back((*phi)[i][qp]);
```

Comment:
should resize on setup and only set here

or try to assert the starting size if you insist on push_back
or check on the size on finalize()

Reply:
Yeah, good catch.